### PR TITLE
new tunnel-based RESP logging and validation

### PR DESCRIPTION
**Pull request by mgravell**
context: https://github.com/StackExchange/StackExchange.Redis/issues/2650

adds:

- opt-in full RESP logging to file (supports TLS, logged data is not encrypted)
- RESP validation of captured files
- RESP local replay of captured files (for display purposes; this is not replay *against a server*)

Usage md: https://github.com/StackExchange/StackExchange.Redis/blob/4b450ce557d52a05af42fbecf3fcc169f0e048ce/docs/RespLogging.md

note that this API is considered (and is annotated) experimental, and as such does not yet feature in the API file list; this is intentional

### CODE REVIEW

File: docs/RespLogging.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+```
+
+If the RESP streams are *not* valid, an exception will provide further details.
+
+**An exception here is strong evidence that there is a fault either in the redis server, or an intermediate proxy**.
+
+Conversely, if the library reported a protocol failure but the validation step here *does not* report an error, then
+that is strong evidence of a library error; [**please report this**](https://github.com/StackExchange/StackExchange.Redis/issues/new) (with details).
+
+You can also *replay* the conversation locally, seeing the individual requests and responses:
```
Comment (philon-msft) [inline docs/RespLogging.md:111]: Could we "replay" files into the library itself for some tests (as opposed to running an actual Redis server locally)?

File: src/StackExchange.Redis/Configuration/LoggingTunnel.cs [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+    public static async Task<long> ReplayAsync(Stream @out, Stream @in, Action<RedisResult, RedisResult> pair)
+    {
+        using Arena<RawResult> arena = new();
+        var outPipe = StreamConnection.GetReader(@out);
+        var inPipe = StreamConnection.GetReader(@in);
+
+        long count = 0;
+        while (true)
+        {
+            var sent = await ReadOneAsync(outPipe, arena, isInbound: false);
```
Comment (philon-msft) [inline src/StackExchange.Redis/Configuration/LoggingTunnel.cs:42]: Append `.ForAwait();` on async calls?

File: docs/RespLogging.md (hunk 1 above)
Comment (mgravell) [reply docs/RespLogging.md:111]: there's a few problems there; we'd need to isolate away anything non-deterministic, which means:

1. anything involving multiple threads, including anything that is a side-effect of heartbeat and (due to timing) anything pub/sub on RESP2
2. anything random, such as the unique muxer identifier (we'd have to make it deterministic in for this scenario, or make it forgive the value received)
3. anything involving multiple connections (cluster)

If we had a concrete scenario where all of the above would be met, and replay would be useful: fine, but I don't think we should do anything preemptively.

File: src/StackExchange.Redis/Configuration/LoggingTunnel.cs (hunk 2 above)
Comment (mgravell) [reply src/StackExchange.Redis/Configuration/LoggingTunnel.cs:42]: good spot; fixed in 64f31748b70998614a0ea06f5c0d07743540254c, although oddly that isn't showing on GitHub yet (despite being pushed)

File: src/StackExchange.Redis/Configuration/LoggingTunnel.cs [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+                break; // no more data, or trailing incomplete messages
+            }
+        }
+        return default;
+    }
+
+    /// <summary>
+    /// Validate a RESP stream and return the number of top-level RESP fragments.
+    /// </summary>
+    public static async Task<long> ValidateAsync(string path)
```
Comment (philon-msft) [inline src/StackExchange.Redis/Configuration/LoggingTunnel.cs:117]: Add <param> doc comment about how you can pass a path to a file OR directory?

Comment (mgravell) [reply src/StackExchange.Redis/Configuration/LoggingTunnel.cs:117]: fair; 5ac13be3a5ad1f06ed725ebd5bbd8fcf18324476

**Review by philon-msft — APPROVED** [review]

**Pull request merged**